tools/flang1/flang1exe/func.c

```diff
@@ -3967,13 +3967,24 @@ rewrite_calls(void)
         }
         sptr_lhs = memsym_of_ast(A_SRCG(ast));
         if (allocatable_member(sptr_lhs)) {
-          rewrite_deallocate(A_SRCG(ast), false, std);
-          if (!ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) &&
-              !POINTERG(sptr_lhs)) {
-            /* Has allocatable members but item itself is not
-             * allocatable nor pointer
-             */
-            nop_dealloc(sptr_lhs, ast);
+          if (!HCCSYMG(sptr_lhs)) {
+            rewrite_deallocate(A_SRCG(ast), false, std);
+            if (!ALLOCG(sptr_lhs) && !ALLOCATTRG(sptr_lhs) &&
+                !POINTERG(sptr_lhs)) {
+              /* Has allocatable members but item itself is not
+              * allocatable nor pointer
+              */
+              nop_dealloc(sptr_lhs, ast);
+            }
+          } else if (HCCSYMG(sptr_lhs) && !ALLOCG(sptr_lhs) &&
+            !ALLOCATTRG(sptr_lhs) && !POINTERG(sptr_lhs)) {
+              /* if sptr_lhs is a temp variable, but it has
+              * allocatable members and item itself is not
+              * allocatable nor pointer.
+              * still rewrite deallocate and perform nop_dealloc
+              */
+              rewrite_deallocate(A_SRCG(ast), false, std);
+              nop_dealloc(sptr_lhs, ast);
```

Comment:
Expanding this block into two cases minimizes changes to the original code, but makes the logic unclear. I suggest checking the allocatable/pointer property only once, and simplify the control flow to call `rewrite_deallocate` and `nop_dealloc` only once each.

Is a compiler-generated temporary allocatable variable guaranteed to be shallow-copied to another variable? If it is, please cite the relevant code in the comments. Otherwise, this code is at risk of leaking memory.

Reply:
I understand that you have cited the code in transfrm.c that transforms allocatable assignment, but it seems that there is no way for the deallocation transformation to know that the temporary variable has been shallow-copied (or in other words, that its elements have escaped to another container).

Also, does your fix need to check the state of `XBIT(54, 0x4)`?

Reply:
1. Thanks for your suggesion, I simplified it based on your comment.
2. I think compiler-generated temporary allocatable variables are guaranteed to be shallow-copied. Here is the reason.

- The shallow-copied operation is performed by `tranform()` function.
- It make a call to `find_allocatable_assignment()`.
- Inside `find_allocatable_assignment()`, `rewrite_allocatable_assignment()` is responsible for rewriting.
- And `handle_allocatable_members()` actually rewrites the ast nodes.
  ```
if (!ALLOCATTRG(sptrdest) || A_TYPEG(astdest) == A_SUBSCR) {
  if (DTYG(dtypedest) == TY_DERIVED && !HCCSYMG(sptrdest) && !XBIT(54, 0x4) &&
      allocatable_member(sptrdest) &&
      /* Stop calling handle_allocatable_members() if the member is an
       * allocatable component of recursive type, otherwise would cause an
       * infinite loop. */
      !handle_recursive_alloc_members) {
    handle_allocatable_members(astdest, astsrc, std, false);
    ast_to_comment(astasgn);
    return;
  }
  if (STYPEG(sptrdest) == ST_MEMBER && !XBIT(54, 0x4) && XBIT(54, 0x1)) {
    /* FS#19118 - this typically occurs with an intrinsic assignment
     * that has a structure constructor on the right hand side. We need
     * to make sure the parent object is allocated when -Mallocatable=03
     * is used.
     */
    astdest = A_PARENTG(astdest);
    if (A_TYPEG(astdest) == A_SUBSCR)
      astdest = A_LOPG(astdest);
    if (A_TYPEG(astdest) == A_MEM) {
      sptrdest = A_SPTRG(A_MEMG(astdest));
    } else
      sptrdest = A_SPTRG(astdest);
    dtypedest = A_DTYPEG(astdest);
    if (!ALLOCATTRG(sptrdest) || DTY(dtypedest) == TY_ARRAY)
      return;
    alloc_scalar_parent_only = TRUE; /* not returning on this one path */
  } else {
    return;
  }
}
```

- The above code snippet contains the first occurrence of `handle_allocatable_members()`.
- Looking closely, anything satisfied `!ALLOCATTRG(sptrdest)` shall fall into this condition.
- According to documentation, only explicitly declared allocatable attribute would has this flag.
- Compiler created temporaries should not have **ALLOCATTR** flag and will fall into this category.

Reply:
3. Rewriting deallocation happens before rewriting allocattion. And it seems no code is used to keep track of temporaries.
4. I don't think there are needed since those xflag are used for allocation part. Not included in the original deallocation code either.

Reply:
Thank you very much for the detailed explanation. Can you summarize your reasoning (in not as many words) in the code comments, or the commit message?

Reply:
> Thanks for your suggesion, I simplified it based on your comment.

Your current branch is still the same as before.